README.md
# ptr-utils

Panther FE utils


CHANGELOG.md
# v1.6.0 (Fri Jun 24 2022)

#### 🚀 Enhancement

- Add auto package [#39](https://github.com/gisat-panther/ptr-utils/pull/39) ([@vdubr](https://github.com/vdubr))
- :sparkles: [#38](https://github.com/gisat-panther/ptr-utils/pull/38) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.5.4 (Mon Jan 24 2022)

#### 🐾 Patch

- Add rollup to externals [#32](https://github.com/gisat-panther/ptr-utils/pull/32) ([@vdubr](https://github.com/vdubr))
- Fix/crypto [#31](https://github.com/gisat-panther/ptr-utils/pull/31) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.5.3 (Thu Jan 20 2022)

#### 🐾 Patch

- Typo, use "Crypto" instead of "crypto" [#30](https://github.com/gisat-panther/ptr-utils/pull/30) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.5.2 (Tue Oct 26 2021)

#### 🐾 Patch

- Map styles: enable attribute transformations for intervals [#29](https://github.com/gisat-panther/ptr-utils/pull/29) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.5.1 (Fri Oct 15 2021)

#### 🐾 Patch

- Fix/dependencies [#28](https://github.com/gisat-panther/ptr-utils/pull/28) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.5.0 (Mon Oct 11 2021)

#### 🚀 Enhancement

- Put react as a peerDependency. [#27](https://github.com/gisat-panther/ptr-utils/pull/27) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.4.0 (Mon Sep 13 2021)

#### 🚀 Enhancement

- Map style utils refactoring & rasters support [#25](https://github.com/gisat-panther/ptr-utils/pull/25) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.3.14 (Wed May 12 2021)

#### 🐾 Patch

- getRemSize returns value on browser [#22](https://github.com/gisat-panther/ptr-utils/pull/22) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.3.13 (Thu Apr 22 2021)

#### 🐾 Patch

- Feature/build improvements [#19](https://github.com/gisat-panther/ptr-utils/pull/19) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.3.12 (Fri Feb 19 2021)

#### 🐾 Patch

- Opt/size tests [#18](https://github.com/gisat-panther/ptr-utils/pull/18) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.3.11 (Mon Feb 15 2021)

#### 🐾 Patch

- Dependencies/upgrade [#17](https://github.com/gisat-panther/ptr-utils/pull/17) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.3.10 (Mon Feb 15 2021)

#### 🐾 Patch

- Bump ini from 1.3.5 to 1.3.8 [#14](https://github.com/gisat-panther/ptr-utils/pull/14) ([@dependabot[bot]](https://github.com/dependabot[bot]))

#### Authors: 1

- [@dependabot[bot]](https://github.com/dependabot[bot])

---

# v1.3.9 (Thu Feb 11 2021)

#### 🐾 Patch

- Projections [#16](https://github.com/gisat-panther/ptr-utils/pull/16) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.3.8 (Thu Feb 11 2021)

#### 🐾 Patch

- Dependencies, test fixes & prettify [#15](https://github.com/gisat-panther/ptr-utils/pull/15) ([@vlach1989](https://github.com/vlach1989))

#### Authors: 1

- Pavel Vlach ([@vlach1989](https://github.com/vlach1989))

---

# v1.3.7 (Thu Sep 24 2020)

#### 🐾 Patch

- Add check, if rooup build does not contains some unresolved external … [#13](https://github.com/gisat-panther/ptr-utils/pull/13) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.3.6 (Fri Sep 11 2020)

#### 🐾 Patch

- Import more maps utils [#12](https://github.com/gisat-panther/ptr-utils/pull/12) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.3.5 (Thu Sep 10 2020)

#### 🐾 Patch

- Import utils functionality from ptr-maps [#11](https://github.com/gisat-panther/ptr-utils/pull/11) ([@vdubr](https://github.com/vdubr))

#### Authors: 1

- Vojtěch Dubrovský ([@vdubr](https://github.com/vdubr))

---

# v1.3.4 (Tue Aug 18 2020)

#### 🐾 Patch

- Bump elliptic from 6.5.2 to 6.5.3 [#10](https://github.com/gisat-panther/ptr-utils/pull/10) ([@dependabot[bot]](https://github.com/dependabot[bot]))

#### Authors: 1

- [@dependabot[bot]](https://github.com/dependabot[bot])

---

# v1.3.3 (Tue Jul 21 2020)

#### 🐾 Patch

- Bump lodash from 4.17.15 to 4.17.19 [#9](https://github.com/gisat-panther/ptr-utils/pull/9) ([@dependabot[bot]](https://github.com/dependabot[bot]))

#### Authors: 1

- [@dependabot[bot]](https://github.com/dependabot[bot])

---

# v1.3.2 (Wed Jul 15 2020)

#### 🐾 Patch

- Fix release [#8](https://github.com/gisat-panther/ptr-utils/pull/8) ([@nenadalm](https://github.com/nenadalm))

#### Authors: 1

- Miloslav Nenadál ([@nenadalm](https://github.com/nenadalm))

---

# v1.3.1 (Thu Jul 09 2020)

#### 🐾 Patch

- Add auto to make automatic releases [#7](https://github.com/gisat-panther/ptr-utils/pull/7) ([@nenadalm](https://github.com/nenadalm) [@tmdmc](https://github.com/tmdmc))

#### Authors: 2

- [@tmdmc](https://github.com/tmdmc)
- Miloslav Nenadál ([@nenadalm](https://github.com/nenadalm))


src/utils.js
import {mapValues as _mapValues} from 'lodash';
import period from './period';
import {isObjectLike as _isObjectLike} from 'lodash';
import crypto from 'crypto';

export default {
	period,

	/**
	 * Generates v4 compliant UUID
	 * @returns {string|*|void}
	 */
	uuid() {
		if (typeof crypto?.randomUUID === 'function') {
			return crypto.randomUUID();
		} else if (typeof crypto?.getRandomValues === 'function') {
			return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, c =>
				(
					c ^
					(crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (c / 4)))
				).toString(16)
			);
		} else {
			//fallback for undefined crypto
			let d = new Date().getTime(); //Timestamp
			let d2 =
				(typeof performance !== 'undefined' &&
					performance.now &&
					performance.now() * 1000) ||
				0; //Time in microseconds since page-load or 0 if unsupported
			return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(
				/[xy]/g,
				function (c) {
					var r = Math.random() * 16; //random number between 0 and 16
					if (d > 0) {
						//Use timestamp until depleted
						r = (d + r) % 16 | 0;
						d = Math.floor(d / 16);
					} else {
						//Use microseconds since page-load if supported
						r = (d2 + r) % 16 | 0;
						d2 = Math.floor(d2 / 16);
					}
					return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
				}
			);
		}
	},

	/**
	 * Return true if uuid parameter is valid uuidv4, else return false.
	 * About uuid versions: https://www.uuidtools.com/decode
	 * @param {string} uuid
	 * @returns {Boolean}
	 */
	isValidUuidV4(uuid) {
		let s = '' + uuid;
		s = s.match(
			'^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[4][0-9a-fA-F]{3}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$'
		);
		if (s === null) {
			return false;
		}
		return true;
	},

	/**
	 * Generates random string of specified length.
	 * @param length
	 * @returns {string}
	 */
	randomString: length =>
		((Math.random() * Math.pow(36, length)) >> 0).toString(36),

	getRemSize: () => {
		if (typeof window === 'undefined') {
			//return default fontsize 16px when run on SSR
			return 16;
		} else {
			return parseFloat(getComputedStyle(document.documentElement).fontSize);
		}
	},

	scrollTo(elementId, containerId) {
		let element = document.getElementById(elementId);

		let elementOffset = element.offsetTop;

		document.getElementById(containerId).scrollTop = elementOffset;
	},

	/**
	 * Get json-like formatted string
	 * @param json {JSON | string | null}
	 * @return {string}
	 */
	getStringFromJson(json) {
		if (json && typeof json === 'object') {
			return JSON.stringify(json, null, 2);
		} else {
			return json;
		}
	},

	/**
	 * Takes deep object and returns it with values containing path to that key in the object (where value isn't a nested object).
	 * e.g. {a: null, b: {c: null, d: null}} => {a: 'a', b: {c: 'b.c', d: 'b.d'}}
	 * Used for constants.
	 * @param object - input object or nested object when called recursively
	 * @param path - path to nested object when called recursively
	 * @returns object
	 */
	deepKeyMirror(object, path) {
		if (_isObjectLike(object)) {
			return _mapValues(object, (value, key) => {
				return this.deepKeyMirror(value, path ? path + '.' + key : key);
			});
		} else {
			return path;
		}
	},

	stringToNumHash(string) {
		string = '' + string;
		if (typeof string !== 'string')
			throw new Error('stringToNumHash: argument must be a string');
		let hash = 1;
		for (let i = 0; i < string.length; i++) {
			hash = Math.imul((hash + string.charCodeAt(i)) | 0, 265443576107);
		}
		return (hash ^ (hash >>> 17)) >>> 0;
	},

	/**
	 * Deterministic colour set based on input string.
	 * @param string - input string
	 * @param count - number of colours
	 * @param options - hue, saturation and lightness ranges
	 * @returns {Array} css hsl codes
	 */
	stringToColours(string, count, options) {
		let hash = this.stringToNumHash(string);
		let colours = [];
		let defaults = {
			hue: [0, 360],
			saturation: [35, 65],
			lightness: [40, 60],
		};
		options = {...defaults, ...options};
		let h, s, l;
		let hueRange = options.hue[1] - options.hue[0];
		let saturationRange = options.saturation[1] - options.saturation[0];
		let lightnessRange = options.lightness[1] - options.lightness[0];
		for (let i = 0; i < (count || 1); i++) {
			h = (hash % hueRange) + options.hue[0];
			hash = (hash ^ (hash >>> 17)) >>> 0;
			s = (hash % saturationRange) + options.saturation[0];
			hash = (hash ^ (hash >>> 17)) >>> 0;
			l = (hash % lightnessRange) + options.lightness[0];
			hash = (hash ^ (hash >>> 17)) >>> 0;
			colours.push('hsl(' + h + ',' + s + '%,' + l + '%)');
		}
		return colours;
	},
};
